components/script/dom/eventsource.rs

```diff
@@ -378,3 +574,31 @@ impl Runnable for RefetchRequestRunnable {
         global.core_resource_thread().send(CoreResourceMsg::Fetch(request, action_sender)).unwrap();
     }
 }
+
+pub struct DispatchEventRunnable {
+    event_source: Trusted<EventSource>,
+    origin: String,
+    type_: Atom,
+    data: String,
+}
+
+impl Runnable for DispatchEventRunnable {
+    fn name(&self) -> &'static str { "EventSource DispatchEventRunnable" }
+
+    // https://html.spec.whatwg.org/multipage/#dispatchMessage
+    #[allow(unsafe_code)]
+    fn handler(self: Box<DispatchEventRunnable>) {
+        // Step 4-5
+        let event_source = self.event_source.root();
+        let global = event_source.global();
+        let _ac = JSAutoCompartment::new(global.get_cx(), event_source.reflector().get_jsobject().get());
+
+        rooted!(in(global.get_cx()) let mut data = UndefinedValue());
+        unsafe { self.data.to_jsval(global.get_cx(), data.handle_mut()) };
+        let event = MessageEvent::new(&*global, self.type_.clone(), false, false, data.handle(),
```

Comment:
So the problem of creating this before the runnable is queued is that `MessageEvent` is not `Send`, perhaps I should wrap it inside a `Trusted` or an `Arc`?